### PR TITLE
test: Add new AMPStorage test, also into iOS Tests scheme

### PR DESCRIPTION
**Pull request by dantetam**
### Summary

Add tests for the AMPStorage file, a new file created for the iOS architecture revamp

### Checklist

* [ X ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-iOS/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

File: Sources/Amplitude/AMPStorage.m [hunk 1]
```diff
@@ -38,6 +38,9 @@ + (NSString *)getAppStorageAmpDir:(NSString *)instanceName {
     NSArray *paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
     NSString *path = [paths firstObject];
     NSString *bundleIdentifier = [[NSBundle mainBundle] bundleIdentifier];
+    if (instanceName == NULL || [instanceName length] == 0) {
```
Comment (yuhao900914) [inline Sources/Amplitude/AMPStorage.m:41]: Do we need this? the migration happens always after `instance` or `instanceName`, so we will have the default instance or other customize instances name.

Comment (dantetam) [reply Sources/Amplitude/AMPStorage.m:41]: In my test, the file name was 
`/Users/dante.tam/Library/Developer/CoreSimulator/Devices/00438800-9EF0-409D-A735-D3A9552A16F0/data/Library/Application Support/com.apple.dt.xctest.tool//amplitude_event_storage.txt`

The double slash messes things up. Adding `DEFAULT_INSTANCE` changes the path URL to
`/Users/dante.tam/Library/Developer/CoreSimulator/Devices/00438800-9EF0-409D-A735-D3A9552A16F0/data/Library/Application Support/com.apple.dt.xctest.tool/DEFAULT_INSTANCE/amplitude_event_storage.txt`

File: Sources/Amplitude/AMPStorage.m [hunk 2]
```diff
@@ -38,6 +38,9 @@ + (NSString *)getAppStorageAmpDir:(NSString *)instanceName {
...
+        instanceName = @"DEFAULT_INSTANCE";
```
[the remark this answers is not in the document]
Comment (dantetam) [reply Sources/Amplitude/AMPStorage.m:42]: Yes, we definitely should. Fixed

File: Amplitude.xcodeproj/xcshareddata/xcschemes/Amplitude_iOSTests.xcscheme [hunk 3]
```diff
@@ -22,6 +22,16 @@
                ReferencedContainer = "container:Amplitude.xcodeproj">
             </BuildableReference>
          </TestableReference>
+         <TestableReference
+            skipped = "YES">
+            <BuildableReference
+               BuildableIdentifier = "primary"
```
Comment (qingzhuozhen) [inline Amplitude.xcodeproj/xcshareddata/xcschemes/Amplitude_iOSTests.xcscheme:28]: Adding a new test schema? Was this during the try or intentional? I guess it won't hurt though, just if we have a new schema for test do we also want to config it in github action to test through?

File: Tests/AMPStorageTests.m [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    XCTAssertNotEqual([dir rangeOfString:expectedDir].location, NSNotFound);
+}
+
+- (void)testGetDefaultIdentifyFile {
+    NSString* dir = [AMPStorage getDefaultIdentifyFile:@"INSTANCE_NAME"];
+    NSString* expectedDir = @"/Application Support/com.apple.dt.xctest.tool/INSTANCE_NAME/amplitude_identify_storage.txt";
+    XCTAssertNotEqual([dir rangeOfString:expectedDir].location, NSNotFound);
+
+    dir = [AMPStorage getDefaultIdentifyFile:@""];
+    NSLog(@"%@", dir);
```
Comment (qingzhuozhen) [inline Tests/AMPStorageTests.m:68]: clean up log

File: Amplitude.xcodeproj/xcshareddata/xcschemes/Amplitude_iOSTests.xcscheme (hunk 3 above)
Comment (dantetam) [reply Amplitude.xcodeproj/xcshareddata/xcschemes/Amplitude_iOSTests.xcscheme:28]: Not intentional. Must have been one of my test schemas.

File: Sources/Amplitude/AMPStorage.m [hunk 5]
```diff
@@ -119,18 +123,17 @@ + (NSMutableArray *)getEventsFromDisk:(NSString *)path {
...
+    if (data == nil) {
         NSString *emptyData = @"{ \"batch\": []}";
         data = [emptyData dataUsingEncoding:NSUTF8StringEncoding];
     }
 
-    NSError *err = NULL;
+    NSError *err = nil;
     NSDictionary *jsonAsDict = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&err];
-    if (err != NULL) {
+    if (err != nil || jsonAsDict == nil) {
```
Comment (dantetam) [inline Sources/Amplitude/AMPStorage.m:134]: If anyone is wondering about this part, invalid json will error silently and then *jsonAsDict will be a nil pointer.

Comment (dantetam) [reply Sources/Amplitude/AMPStorage.m:134]: Fail silently meaning that *err will also be nil

File: Tests/AMPStorageTests.m [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    XCTAssertTrue([fileManager fileExistsAtPath:customPath]);
+    
+    [AMPStorage finish:customPath];
+    
+    NSString *content = [NSString stringWithContentsOfFile:customPath encoding:NSUTF8StringEncoding error:nil];
+    XCTAssertNotEqual([content rangeOfString:exampleEvent].location, NSNotFound);
+    XCTAssertNotEqual([content rangeOfString:@"{ \"batch\":"].location, NSNotFound);
+    
+    NSMutableArray *finishedFileDict = [AMPStorage getEventsFromDisk:customPath];
+    NSDictionary *secondEvent = finishedFileDict[1];
```
Comment (qingzhuozhen) [inline Tests/AMPStorageTests.m:120]: maybe we can add an assert for the finishedFileDict count first

Comment (dantetam) [reply Tests/AMPStorageTests.m:120]: That's a good idea, will do.

File: Sources/Amplitude/AMPStorage.m (hunk 5 above)
Comment (qingzhuozhen) [reply Sources/Amplitude/AMPStorage.m:134]: Good catch! I was hoping error is not nil if it is invalid json

File: Tests/AMPStorageTests.m [hunk 7]
```diff
@@ -0,0 +1,135 @@
+//
+//  AMPStorageTests.m
+//
+//
+//  Created by Dante Tam on 7/22/21.
+//  Copyright © 2018 Amplitude. All rights reserved.
```
Comment (qingzhuozhen) [inline Tests/AMPStorageTests.m:6]: Can we remove the date, and put 2021 for copy right?

File: Tests/AMPStorageTests.m [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    
+    NSMutableArray *finishedFileDict = [AMPStorage getEventsFromDisk:customPath];
+    NSDictionary *secondEvent = finishedFileDict[1];
+    XCTAssertEqualObjects([secondEvent objectForKey:@"event_type"], @"testEvent");
+    
+    //Test a file that has not been finished
+    [AMPStorage remove:customPath];
+    XCTAssertFalse([fileManager fileExistsAtPath:customPath]);
+    
+    [AMPStorage storeEventAtUrl:url event:exampleEvent];
```
Comment (qingzhuozhen) [inline Tests/AMPStorageTests.m:127]: Curious why we are test again here? Test the url can be reused again?

File: Sources/Amplitude/AMPStorage.m [hunk 9]
```diff
@@ -24,10 +24,11 @@
 #import <Foundation/Foundation.h>
 #import <Amplitude.h>
 #import "AMPStorage.h"
+#import "AMPConstants.h"
 
 @implementation AMPStorage
 
-+ (BOOL)hasFileStorage:(NSString *) instanceName {
++ (BOOL)hasFileStorage:(NSString *)instanceName {
```
Comment (qingzhuozhen) [inline Sources/Amplitude/AMPStorage.m:31]: This method seems not covered in the tests, can we cover it as well?

File: Tests/AMPStorageTests.m (hunk 8 above)
Comment (dantetam) [reply Tests/AMPStorageTests.m:127]: It's testing the same filename rewritten again, but this time it does not have the method `finish` called on it.

Comment (qingzhuozhen) [reply Tests/AMPStorageTests.m:134]: Gotcha, sounds fine for me

**Review by qingzhuozhen — APPROVED** [review]

File: Tests/AMPStorageTests.m [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+#import <XCTest/XCTest.h>
+#import "AMPStorage.h"
+
+@interface AMPStorageTests : XCTestCase
+    
+@end
+
+@implementation AMPStorageTests {}
+
+NSString *exampleEvent = @"{\n    \"api_properties\" :     {\n        \"ios_idfv\" : \"2D623DDA-4AC8-4A1E-BB4B-8264CB2AB31A\"\n    },\n    \"carrier\" : \"Unknown\",\n    \"country\" : \"United States\",\n    \"device_id\" : \"8C027883-5AA1-43F8-8FB8-7837EBE519C4\",\n    \"event_type\" : \"testEvent\",\n \"user_id\" : \"userId\",\n    \"user_properties\" :     {\n    },\n    \"uuid\" : \"41EEDEBB-3A7E-4EDF-BEFC-689D7912BCF3\"\n}";
```
Comment (yuhao900914) [inline Tests/AMPStorageTests.m:18]: Not a blocker, do we need to add the `\n` and `    ` in the example?

File: Tests/AMPStorageTests.m [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    NSString* dir = [AMPStorage getAppStorageAmpDir:@""];
+    NSFileManager *fileMgr = [NSFileManager defaultManager];
+    NSArray *fileArray = [fileMgr contentsOfDirectoryAtPath:dir error:nil];
+    for (NSString *filename in fileArray)  {
+        [fileMgr removeItemAtPath:[dir stringByAppendingPathComponent:filename] error:NULL];
+    }
+}
+
+- (void)testHasFileStorage {
+    BOOL hasFileStorage = [AMPStorage hasFileStorage:@"INSTANCE_NAME"];
```
Comment (yuhao900914) [inline Tests/AMPStorageTests.m:35]: In our code base, the instance name should be always lowercase. Do you think we need to make it consistant?

File: Tests/AMPStorageTests.m [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    
+    dir = [AMPStorage getAppStorageAmpDir:NULL];
+    expectedDir = @"/Application Support/com.apple.dt.xctest.tool/";
+    XCTAssertNotEqual([dir rangeOfString:expectedDir].location, NSNotFound);
+}
+
+- (void)testGetDefaultEventsFile {
+    NSString* dir = [AMPStorage getDefaultEventsFile:@"INSTANCE_NAME"];
+    NSString* expectedDir = @"/Application Support/com.apple.dt.xctest.tool/INSTANCE_NAME/amplitude_event_storage.txt";
+    XCTAssertNotEqual([dir rangeOfString:expectedDir].location, NSNotFound);
```
Comment (yuhao900914) [inline Tests/AMPStorageTests.m:56]: If there is no event stored in file storage yet, will this case be broken?

File: Tests/AMPStorageTests.m [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    [super setUp];
+}
+
+- (void)tearDown {
+    [super tearDown];
+    NSString* dir = [AMPStorage getAppStorageAmpDir:@""];
+    NSFileManager *fileMgr = [NSFileManager defaultManager];
+    NSArray *fileArray = [fileMgr contentsOfDirectoryAtPath:dir error:nil];
+    for (NSString *filename in fileArray)  {
+        [fileMgr removeItemAtPath:[dir stringByAppendingPathComponent:filename] error:NULL];
```
Comment (yuhao900914) [inline Tests/AMPStorageTests.m:30]: Does this only remove the file in the default instance folder, do you think we need to do anything for other instances we created in the test?

File: Tests/AMPStorageTests.m (hunk 10 above)
Comment (dantetam) [reply Tests/AMPStorageTests.m:18]: This is an actual event generated by the SDK that I printed out. That's why the data is like this. It's closest to the real world use.

File: Tests/AMPStorageTests.m (hunk 11 above)
Comment (dantetam) [reply Tests/AMPStorageTests.m:35]: It's just for the test. Instance name here is just to match with the string later

File: Tests/AMPStorageTests.m (hunk 13 above)
Comment (dantetam) [reply Tests/AMPStorageTests.m:30]: This should remove every file and directory under `.../Application Support/com.apple.dt.xctest.tool/`

Comment (dantetam) [reply Tests/AMPStorageTests.m:30]: What else should we do here @yuhao900914 ? Let me know if I'm missing something.

Comment (yuhao900914) [reply Tests/AMPStorageTests.m:30]: Oh my bad, yeah, this should works fine.

**Review by yuhao900914 — APPROVED** [review]

**Comment by dantetam** [comment]
Again, thank you @yuhao900914 and @qingzhuozhen for the useful input.

**Pull request merged**